sheepdog/auth/__init__.py

```diff
@@ -39,3 +42,88 @@ def _log_import_error(module_name):
     from authutils import require_auth
 except ImportError:
     _log_import_error('require_auth')
+
+
+def _role_error_msg(user_name, roles, project):
+    role_names = [
+        role if role != '_member_' else 'read (_member_)' for role in roles
+    ]
+    return (
+        "User {} doesn't have {} access in {}".format(
+            user_name, ' or '.join(role_names), project
+        )
+    )
+
+
+def get_program_project_roles(program, project):
+    """
+    A lot of places (submission entities etc.) confuse the terminology and have
+    a ``project_id`` attribute which is actually ``{program}-{project}``, so
+    in those places call this function like
+
+        get_program_project_roles(*project_id.split('-'))
+
+    Args:
+        program (str): program name (NOT id)
+        project (str): project name (NOT id)
+
+    Return:
+        Set[str]: roles
+    """
+    # Get the actual CurrentUser instance behind the werkzeug proxy so we can
+    # slap this attributes on it
+    user = current_user._get_current_object()
+
+    if not hasattr(user, 'sheepdog_roles'):
+        user.sheepdog_roles = dict()
```

Comment:
I feel it might be more clear to put this on sheepdog flask.g instead of back to authutils' user object?

Reply:
good idea, this is kind of a hack